### PR TITLE
Smart parallelization

### PR DESCRIPTION
**Pull request by llxia**
- Generate and trigger child test jobs based on PARALLEL value
- PARALLEL value can be the following:
   - Subdir (i.e., this is for external, perf, special.system test builds)
   - Dynamic
      - set number of child jobs (childJobNum). childJobNum =
params.NUM_MACHINES. If it is not provided, the default value is 1.
childJobNum cannot be greater than slave machines / 2
      - call TKG with NUM_MACHINES=childJobNum. It will generate
parallelList.mk dynamically
      - get NUM_LIST from parallelList.mk. Set childJobNum = NUM_LIST
   - Iteration (simply repeat based on ITERATIONS param. Max is 20)
   - None (no parallel)
- template is updated to create test jobs accordingly

Issue: #1563

Signed-off-by: lanxia <lan_xia@ca.ibm.com>

### CODE REVIEW

**Comment by llxia** [comment]
|  PARALLEL | TARGET  | Grinder |
|:-:|:-:|:-:|
|  Iteration | sanity.openjdk  | Grinder_Sandbox/140 |
|  Dynamic | sanity.openjdk  | Grinder_Sandbox/142  |
|  Subdir | sanity.system  | Grinder_Sandbox/139  |
|  None | jdk_math_j9  | Grinder_Sandbox/141  |

**Comment by llxia** [comment]
Blocked by runtimes/infrastructure/issues/3640

**Comment by llxia** [comment]
- This PR needs to be merged together with https://github.com/eclipse/openj9/pull/9388
- Due to the template change in this PR, it needs to be approved in openj9 Jenkins and AdoptOpenJDK Jenkins

**Comment by llxia** [comment]
Revert the change of making `ADOPTOPENJDK_REPO ` and `ADOPTOPENJDK_BRANCH ` as variable in buildenv/jenkins/testJobTemplate, so that this PR will not get blocked by 
runtimes/infrastructure/issues/3640

Please note after reverting, test builds will always git clone https://github.com/AdoptOpenJDK/openjdk-tests.git master branch regardless `ADOPTOPENJDK_REPO ` and `ADOPTOPENJDK_BRANCH ` value.

**Comment by llxia** [comment]
- retest:
jdk_math_j9 Grinder_Sandbox/147
dynamic parallel Grinder_Sandbox/148

- test template:
Test_Job_Auto_Gen_Sandbox/30
Test_openjdk8_j9_sanity.functional_x86-64_mac_Nightly/377/

**Comment by smlambert** [comment]
Re: https://github.com/AdoptOpenJDK/openjdk-tests/pull/1758#issuecomment-621315554 that seems an acceptable compromise for now, as use case for auto-generated test builds would be to likely always do that.  Grinder which is not auto-generated is where users get to employ that option (of using their own fork/branch).

**Comment by llxia** [comment]
PR is updated and ready for review.

File: buildenv/jenkins/JenkinsfileBase [hunk 1]
```diff
@@ -108,52 +108,102 @@ def setupEnv() {
...
+						archiveArtifacts artifacts: 'parallelList.mk', fingerprint: true, allowEmptyArchive: false
+					}
 
-			//Setup for option-specific parameters
-			//If ITERATIONS is 1, then parallel is based on subfolders
-			if (repeatAmount == 1) {
+					UPSTREAM_TEST_JOB_NAME = JOB_NAME
+					UPSTREAM_TEST_JOB_NUMBER = BUILD_NUMBER
+				} else {
+					assert false : "Build failed becuase cannot find NUM_LIST in parallelList.mk file."
```
Comment (smlambert) [inline buildenv/jenkins/JenkinsfileBase:148]: spelling: becuase -> because

File: buildenv/jenkins/testJobTemplate [hunk 2]
```diff
@@ -171,6 +183,8 @@ ARCH_OS_LIST.each { ARCH_OS ->
 							stringParam('TKG_REPO', TKG_REPO, "TestKitGen git repo. Please use ssh for zos.")
 							stringParam('TKG_BRANCH', "master", "TKG branch")
 							stringParam('TKG_SHA', "", "TKG sha")
+							stringParam('UPSTREAM_TEST_JOB_NAME', "", "Auto-populated. Upstream test job name. It will be used togehter with PARALLEL=Dynamic")
```
Comment (smlambert) [inline buildenv/jenkins/testJobTemplate:186]: spelling togehter -> together, same below

**Comment by llxia** [comment]
PR is updated

**Comment by llxia** [comment]
Eclipse Bugzilla: https://bugs.eclipse.org/bugs/show_bug.cgi?id=562706

**Pull request merged**